bench.c

```diff
@@ -0,0 +1,189 @@
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <pthread.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <time.h>
+#include <unistd.h>
+
+#define TARGET_HOST "127.0.0.1"
+#define TARGET_PORT 12345
+#define BENCH_COUNT 10
+#define RESULT_FILE_NAME "bench.txt"
+
+/* vary-length message should shorter than this */
+#define MAX_MSG_LEN 32
+
+/*
+ * Too much concurrent connection would be treat as sort of DDoS attack
+ * (mainly caused by configs (kernel: `tcp_max_syn_backlog` and
+ * `somaxconn`. Application (fastecho): `backlog`). Nevertheless, default
+ * maximum number of fd per-process is `1024`, if you
+ * insist to proceed the benchmark, perform following modifications:
+ * 
+ * (1)
+ * Use following commands to adjust kernel attributes:
+ *     a. `$ sudo sysctl net.core.somaxconn={depends_on_MAX_THREAD}`
+ *     b. `$ sudo sysctl net.ipv4.tcp_max_syn_backlog={ditto}`
+ * Note that `$ sudo sysctl net.core.somaxconn` can get current value.
+ * `somaxconn` is max amount of established connection, whereas
+ * `tcp_max_syn_backlog` is max amount of connection at first step
+ * of TCP 3-way handshake (SYN).
+ *
+ * (2)
+ * Use command `$ ulimit -n {ditto}` to enlarge limitation of fd
+ * per-process. Note that this modification only effects on process
+ * which executes the command and its child processes.
+ *
+ * (3)
+ * Specify `backlog` with value as large as `net.ipv4.tcp_max_syn_backlog`.
+ *
+ * Remember to reset the modifications after benchmarking to keep
+ * stability of running machine
+ */
+#define MAX_THREAD 1000
+
+/*
+ * TODO: provide unique message (maybe generate dynamically, or somehow) for
+ * each worker could produce benchmark result which is more conforms to 
+ * realworld usage.
+ */
+const char *msg_dum = "dummy message";
+
+pthread_t pt[MAX_THREAD];
+
+/* block all worker before they are all ready to bombing kecho */
+bool ready;
+
+pthread_mutex_t res_lock = PTHREAD_MUTEX_INITIALIZER;
+pthread_mutex_t worker_lock = PTHREAD_MUTEX_INITIALIZER;
+pthread_cond_t worker_wait = PTHREAD_COND_INITIALIZER;
+
+long time_res[MAX_THREAD] = {0};
+int idx = 0; /* for indexing `time_res` */
+FILE *bench_fd;
+
+static inline long time_diff_us(struct timeval *start, struct timeval *end)
+{
+    return ((end->tv_sec - start->tv_sec) * 1000000) +
+           (end->tv_usec - start->tv_usec);
+}
+
+static void *bomb_worker(void *arg)
+{
+    int sock_fd, rt;
+    char dummy[MAX_MSG_LEN];
+    struct timeval start, end;
+
+    /* wait until all worker created */
+    pthread_mutex_lock(&worker_lock);
+    while (!ready)
+        if ((rt = pthread_cond_wait(&worker_wait, &worker_lock))) {
+            puts("cond_wait failed!");
```

Comment:
Unify with `perror` or something alike.

Reply:
Since pthread doesn't cooperate with `perror`, I unify it with `goto` style which also used by other error handler in the worker func.